modules/utxo-lib/src/bitgo/UtxoPsbt.ts

```diff
@@ -0,0 +1,319 @@
+import { UtxoTransaction } from './UtxoTransaction';
+import { taproot, HDSigner, Psbt, PsbtTransaction, Transaction } from '../';
+import { TapBip32Derivation, Transaction as ITransaction, TransactionFromBuffer } from 'bip174/src/lib/interfaces';
+import { checkForInput } from 'bip174/src/lib/utils';
+import { Psbt as PsbtBase } from 'bip174';
+import { Network } from '..';
+import { ecc as eccLib, script as bscript } from '..';
+import * as opcodes from 'bitcoin-ops';
+import { BufferWriter, varuint } from 'bitcoinjs-lib/src/bufferutils';
+
+export interface HDTaprootSigner extends HDSigner {
+  /**
+   * The path string must match /^m(\/\d+'?)+$/
+   * ex. m/44'/0'/0'/1/23 levels with ' must be hard derivations
+   */
+  derivePath(path: string): HDTaprootSigner;
+  /**
+   * Input hash (the "message digest") for the signature algorithm
+   * Return a 64 byte signature (32 byte r and 32 byte s in that order)
+   */
+  signSchnorr(hash: Buffer): Buffer;
+}
+
+export interface SchnorrSigner {
+  publicKey: Buffer;
+  signSchnorr(hash: Buffer): Buffer;
+}
+
+export interface TaprootSigner {
+  leafHashes: Buffer[];
+  signer: SchnorrSigner;
+}
+
+export interface PsbtOpts {
+  network: Network;
+}
+
+export class UtxoPsbt<Tx extends UtxoTransaction<bigint>> extends Psbt {
+  protected static transactionFromBuffer(buffer: Buffer, network: Network): UtxoTransaction<bigint> {
+    return UtxoTransaction.fromBuffer<bigint>(buffer, false, 'bigint', network);
+  }
+
+  static createPsbt(opts: PsbtOpts, data: PsbtBase): UtxoPsbt<UtxoTransaction<bigint>> {
+    return new UtxoPsbt<UtxoTransaction<bigint>>(
+      opts,
+      data || new PsbtBase(new PsbtTransaction({ tx: new UtxoTransaction<bigint>(opts.network) }))
+    );
+  }
+
+  static fromBuffer(buffer: Buffer, opts: PsbtOpts): UtxoPsbt<UtxoTransaction<bigint>> {
+    const transactionFromBuffer: TransactionFromBuffer = (buffer: Buffer): ITransaction => {
+      const tx = this.transactionFromBuffer(buffer, opts.network);
+      return new PsbtTransaction({ tx });
+    };
+    const psbtBase = PsbtBase.fromBuffer(buffer, transactionFromBuffer);
+    const psbt = this.createPsbt(opts, psbtBase);
+    // Upstream checks for duplicate inputs here, but it seems to be of dubious value.
+    return psbt;
+  }
+
+  protected static newTransaction(network: Network): UtxoTransaction<bigint> {
+    return new UtxoTransaction<bigint>(network);
+  }
+
+  get tx(): Tx {
+    return (this.data.globalMap.unsignedTx as PsbtTransaction).tx as Tx;
```

Comment:
Why is the casting necessary here?

Reply:
Because the `unsignedTx` is an `bip174.Transaction` interface, but we need to tell it what concrete implementation `bitcoinjs-lib.PsbtTransaction` we were using to get the `tx` off of it.